src/core/backend/pearl/core.rs

```diff
@@ -128,32 +156,32 @@ impl BackendStorage for Pearl {
 
     async fn shutdown(&self) {
         use futures::stream::FuturesUnordered;
+        use std::iter::once;
         info!("begin shutdown");
-        let futures = FuturesUnordered::new();
-        for dc in self.disk_controllers.iter() {
-            futures.push(async move { dc.shutdown().await })
-        }
-        let _ = futures.collect::<()>().await;
-        // TODO: process alien disk with other ones
-        self.alien_disk_controller.shutdown().await;
+        let futures = self
+            .disk_controllers
+            .iter()
+            .chain(once(&self.alien_disk_controller))
+            .map(|dc| async move { dc.shutdown().await })
+            .collect::<FuturesUnordered<_>>();
+        futures.collect::<()>().await;
         info!("shutting down done");
     }
 
-    async fn blobs_count(&self) -> (usize, usize) {
-        let mut cnt = 0;
-        for dc in self.disk_controllers.iter() {
-            cnt += dc.blobs_count().await
-        }
-        let alien_cnt = self.alien_disk_controller.blobs_count().await;
-        (cnt, alien_cnt)
+    fn disk_controllers(&self) -> Option<(&[Arc<DiskController>], Arc<DiskController>)> {
+        Some((&self.disk_controllers, self.alien_disk_controller.clone()))
     }
 
-    async fn index_memory(&self) -> usize {
-        let mut cnt = 0;
+    async fn close_unneeded_active_blobs(&self, soft: usize, hard: usize) {
         for dc in self.disk_controllers.iter() {
-            cnt += dc.index_memory().await
+            dc.close_unneeded_active_blobs(soft, hard).await;
         }
-        cnt += self.alien_disk_controller.index_memory().await;
-        cnt
+    }
+
+    async fn collect_metrics(&self) -> BackendMetrics {
```

Comment:
Are you sure it is the good way to collect all metrics at once? Because if we need one metric, we still will block at least three times more.

Reply:
I've tried to reduce number of functions in `BackendStorage` trait, because it becomes too bloated. But you're right, that the decision to place all metrics in one function call is weird, so I decided to separate metrics functions with another trait, and require it in `Backend` struct in `inner` field (I don't know about dependencies: should `BackendStorage` trait depend on `MetricsProducer` or not or vice versa, so I decided to make the independent)